KAFKA-2480: Add backoff timeout and support rewinds

Ishiihara opened the pull request.

Ishiihara committed:
Add backoff timeout and support rewinds

ewencp commented on copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:
```diff
@@ -59,6 +63,40 @@ public void offset(Map<TopicPartition, Long> offsets) {
     }
 
     /**
+     * Set the backoff timeout. SinkTasks should use this to indicate that they need to retry certain
```
This might need a bit more explanation. This isn't a guaranteed backoff, right? Generally when I think of backoff, I think nothing is going to happen during that period. But the task may still receive new records. Instead, this is guaranteeing that you will get another callback (possibly with no records) at the latest in backoffMs milliseconds.

We might consider other names for this since the current name isn't really clear. Given that the implementation literally uses this value to cap the timeout, maybe `timeout` would be more appropriate? or `putTimeout` to indicate that it is tied to the `put()` calls?

gwenshap replied:
I think I get what you mean, but this is incredibly ambiguous and if someone didn't work on a connector before, they may not get it at all.

If I understand correctly, what you are saying is:
"As a connector developer, you may have certain operations on external systems that may fail and you may want to retry (for example, an update on a database table may fail due to locks). You will need to know how long to wait until retrying. Please use this method to know how long to wait instead of creating your own configuration. This will give some standardization in our system".

Did I even get it right? If not, we need better explanation for sure :)

Ishiihara replied:
@gwenshap Good explanation. That is exactly the intention. I will add your explanation to the java doc.

ewencp replied:
Wait, that seems like the exact opposite of what we want. This method allows tasks to indicate to the framework the maximum amount of time that should be allowed to pass before their `put()` method is invoked again, regardless of whether there are new records available. The first half of @gwenshap's explanation makes sense, but the second half isn't right -- the idea is to use this to let the task inform the framework about reasonable times, not the other way around.

@Ishiihara This is also why I was suggesting we might want to be very clear about separating the interface for `SinkTaskContext` from the implementation. The interface should only support this method `void backoff(long backoffMs)` to allow tasks to set the maximum delay they will see before the next `put()` call. Then the implementation can include `long backoff()`, which is how the framework gets access to that value.

ewencp commented on copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:
```diff
@@ -59,6 +63,40 @@ public void offset(Map<TopicPartition, Long> offsets) {
+     * operations after the backoff timeout.
+     * @param backoffMs the backoff timeout in milliseconds.
+     */
+    public void backoffMs(long backoffMs) {
```
Let's name these methods so they sound like actions to be taken by the SinkTask. So instead of `public void backoffMs(long backoffMs)`, I think it should probably be `public void backoff(long backoffMs)`. A bit of a nit, but since this is public (and very permanent) API, I'd like to get the naming right.

ewencp commented on copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:
```diff
@@ -59,6 +63,40 @@ public void offset(Map<TopicPartition, Long> offsets) {
+     * Get the backoff in milliseconds set by SinkTasks. Used by the Copycat framework.
+     * @return the backoff timeout in milliseconds.
+     */
+    public long backoffMs() {
```
I'm noticing that we're now including quite a few things in this interface that are really internal details -- the methods that set these values are intended to be public APIs called by SinkTasks, but we also expose methods to get those values back out so the framework can get at them. Should we split this between an interface/abstract class `SinkTaskContext` which is in `copycat/api` and an implementation in `copycat/runtime`?

ewencp commented on copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:
```diff
@@ -59,6 +63,40 @@ public void offset(Map<TopicPartition, Long> offsets) {
+     * partitions needs to rewind.
+     * @param rewinds The set of topic partitions that needs to rewind.
+     */
+    public void rewinds(Set<TopicPartition> rewinds) {
```
Same here -- should probably just be `public void rewind(Set<TopicPartition> partitions)`.

Also, is there a reason we need both `offsets()` and `rewinds()`? Seems like they accomplish basically the same thing, but with slightly different semantics. Why not have just one mechanism for requesting offset changes to the framework, and on each iteration, the `WorkerSinkTask` can handle checking if there are offsets that need to be reset, reset them, and then clear out that list?

The only benefit I see right now is that by having one of them be stateful, you've moved a bit of the work of maintaining the offsets to the framework, but I'm not sure that small benefit to the task developer is worth having nearly duplicate functionality.

Ishiihara committed:
Address review comments

Ishiihara commented:
@ewencp Addressed most review comments. One question is what is the best name for backoff. Currently, at the HDFS connector side, the failure time is recorded and we only proceed to actual retry when the backoff is reached.

ewencp commented on copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/WorkerSinkTask.java:
```diff
@@ -108,6 +108,9 @@ public void close() {
         try {
+            rewind();
+            long backOffMs = context.backoff();
+            timeoutMs = Math.max(timeoutMs, backOffMs);
```
Shouldn't this be `Math.min`? The timeout passed into this poll method includes at least the next commit time. If you use `Math.max`, won't we miss that timeout if the `backoffMs` value is too large? (or vice versa)

ewencp replied:
Also, don't we either need to record the starting time for the backoff or reset it at some point? Otherwise we'll keep using the timeout indefinitely. Or was the expectation that the task reset it when it no longer needs it?

Ishiihara replied:
I think `Math.min` makes more sense now since the connector is actually not proceed unless the backoff is reached. Also updated the PR to reset the backoff.

ewencp commented:
@Ishiihara Not sure about the name. Maybe `backoff` works. Could also be something like `timeout` or `schedule`. `Timeout` might be better as `schedule` isn't really ever associated with spurious wakeups, whereas `timeouts` sometimes are.

Ishiihara committed:
Address review comments

Ishiihara commented:
@ewencp Added backoff reset.

ewencp commented:
@gwenshap Pretty small, but adds a few key features for sink connectors and looks ok to me. Care to take a look? If you have opinions about naming, see discussion of `backoff` vs other terms above.

gwenshap commented on copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTask.java:
```diff
@@ -62,4 +62,11 @@ public void initialize(SinkTaskContext context) {
     public abstract void flush(Map<TopicPartition, OffsetAndMetadata> offsets);
+
+
+    public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
```
undocumented public API? Connector developers will need to know what to implement here.

Ishiihara committed:
Address review comments and add necessary java docs

Ishiihara commented:
@ewencp @gwenshap Just updated the related doc.

Ishiihara committed:
Move some methods from SinkTaskContext to WorkerSinkTaskContext

Ishiihara commented:
@ewencp @gwenshap Refactor the code to move some methods to `WorkerSInkTaskContext` in `runtime`.  The connector will only be able to access the setter methods.

Ishiihara committed:
Change backoff to timeout

gwenshap commented:
LGTM
